File: deepX/IR.rs
use serde::{Serialize, Deserialize};
use std::collections::HashMap;

type TensorId = usize;
type NodeId = usize;

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
enum DataType {
    Int32,
    Int64,
    Float32,
    Float64,
    Bool
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
enum TensorType {
    Input, // tensors from feed_dict
    Output, // output tensor
    Intermediate, // tensors from intermediate computations
    Parameter // learnable parameter tensors
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
enum OpAttr {
    Int(i64),
    Float(f64),
    IntList(Vec<i64>),
    FloatList(Vec<f64>),
    String(String),
    Bool(bool)
}

// Tensor IR
#[derive(Serialize, Deserialize, Debug, Clone)]
struct Tensor {
    id: TensorId,
    shape: Vec<usize>,
    requires_grad: bool,
    tensor_type: TensorType
}

// Node IR
#[derive(Serialize, Deserialize, Debug, Clone)]
struct Node {
    id: NodeId,
    inputs: Vec<TensorId>,
    output: TensorId,
    op_name: String,
    op_attrs: HashMap<String, OpAttr> // kwargs
}

// Graph IR
#[derive(Serialize, Deserialize, Debug, Clone)]
struct Graph {
    version: String,
    tensors: HashMap<TensorId, Tensor>,
    nodes: Vec<Node>,
    inputs: Vec<TensorId>,
    output: TensorId,
    topo_order: Vec<NodeId> // execution order
}

// Compilation Configuration
#[derive(Serialize, Deserialize, Debug, Clone)]
struct CompilationConfig {
    dtype: DataType,
    mixed_precision: bool
}

File: deepX/IR.cpp
#pragma once
#include <cstdint>
#include <string>
#include <vector>
#include <variant>
#include <unordered_map>
#include <nlohmann/json.hpp>

namespace deepx {

using TensorId = std::size_t;
using NodeId   = std::size_t;

enum class DataType { Int32, Int64, Float32, Float64, Bool };
enum class TensorType { Input, Output, Intermediate, Parameter };

using OpAttr = std::variant<
    std::int64_t,
    double,
    std::vector<std::int64_t>,
    std::vector<double>,
    std::string,
    bool>;

struct Tensor {
    TensorId id;
    std::vector<std::size_t> shape;
    bool requires_grad;
    TensorType tensor_type;
};

struct Node {
    NodeId id;
    std::vector<TensorId> inputs;
    TensorId output;
    std::string op_name;
    std::unordered_map<std::string, OpAttr> op_attrs;
};

struct Graph {
    std::string version;
    std::unordered_map<TensorId, Tensor> tensors;
    std::vector<Node> nodes;
    std::vector<TensorId> inputs;
    TensorId output;
    std::vector<NodeId> topo_order;

    static Graph from_json(const nlohmann::json& j);
    nlohmann::json to_json() const;
};

struct CompilationConfig {
    DataType dtype;
    bool mixed_precision;
};

} // namespace deepx
